### PR TITLE
Deposit

### PR DESCRIPTION
**Pull request by bh2smith**
Include basic deposit functionality with units tests.

**Note that** deposit hashed are kept as an array for the purpose of implementing rollbacks in the future.

### CODE REVIEW

File: contracts/BatchAuction.sol [hunk 1]
```diff
@@ -38,4 +54,32 @@ contract BatchAuction is Ownable {
...
+
+        require(
+            ERC20(tokenAddress).transferFrom(msg.sender, address(this), amount), 
+            "Unsuccessful transfer"
+        );
+        
+        // Increment depositIndex until it matches correct.
+        uint nextDepositIndex = block.number / 20;
+        while (depositIndex != nextDepositIndex) {
+            depositIndex++;
```
Comment (josojo) [inline contracts/BatchAuction.sol]: please use here depositIndex = block.number/20

If there were for 60 blocks no deposits, I would prefer to not do the next 3 deposits split over 3 snarks

**Review by josojo — APPROVED** [review]
very good PR. Thanks.

After fixing the one raised comment, you are ready to merge

File: contracts/BatchAuction.sol (hunk 1 above)
Comment (bh2smith) [reply contracts/BatchAuction.sol]: Done! Great suggestion!

**Review by fleupold — COMMENTED** [review]
Sorry I wasn't able to review this change earlier. I think there is a fundamental problem with the way deposits are modeled here if we want to be able to prove them via snarks.

Currently, the deposit hash is bound in time (20 blocks) but unbound in size (how many deposits are "inside" it). However, the snark to apply any state transition has to take a fixed size input. While it's possible to collect less than that limit (we can pad the remaining values with 0), it won't be possible to ever process more deposits (268M R1CS size limit).
Thus I believe we need to limit the maximum number of deposits per hash.

Would you be able to explain, why you chose to bound each deposit state by time?

**Comment by bh2smith** [comment]
@Felix the Block time model for deposits is meant to ensure deposits occur in batches and that no-inflight deposits could affect the deposit hash while it is trying to be submitted.

**Pull request merged**